dolma-cccc/domain_counter.py

```diff
@@ -0,0 +1,239 @@
+            future_to_file = {executor.submit(process_file, file): file for file in all_files}
+
+            suburl_results = []
+            text_length_results = []
+
+            try:
+                for future in tqdm(as_completed(future_to_file), total=len(future_to_file), desc="Processing files"):
+                    try:
+                        suburl_counter, text_length_counter = future.result()
+                        suburl_results.append(suburl_counter)
+                        text_length_results.append(text_length_counter)
+                    except Exception as e:
+                        print(f"Error processing file {future_to_file[future]}: {e}")
+                    finally:
+                        file_progress.update(1)  # Update the outer file-level progress bar
+
+                wait(future_to_file.keys(), return_when=FIRST_EXCEPTION)
+
+            except KeyboardInterrupt:
+                print("Ctrl+C pressed, terminating...")
+                executor.shutdown(wait=False)
+                sys.exit(1)
+
+    final_suburl_counter = merge_counters(suburl_results)
+    final_text_length_counter = merge_counters(text_length_results)
+
+    sorted_suburls = sorted(final_suburl_counter.items(), key=lambda item: item[1], reverse=True)
+    print("Top 10 SubURLs by frequency:")
+    for suburl, count in sorted_suburls[:10]:
+        print(f"{suburl}: {count}")
+
+    sorted_text_lengths = sorted(final_text_length_counter.items(), key=lambda item: item[1], reverse=True)
+    print("\nTop 10 SubURLs by total text length:")
+    for suburl, total_length in sorted_text_lengths[:10]:
+        print(f"{suburl}: {total_length}")
+
+    return final_suburl_counter, final_text_length_counter
+
+
+def calculate_level(suburl):
+    return suburl.count('/')  # Number of slashes indicates the level
+
+
+def find_parent_url(url):
+    """
+    Returns the parent URL by removing the last segment.
+    For example, 'AA/BB/CC' -> 'AA/BB'
+    """
+    parts = url.rstrip('/').split('/')
+    if len(parts) > 1:
+        return '/'.join(parts[:-1])
+    return None
+
+def remove_redundant_urls(dataset):
+    """
+    Remove redundant URLs from the dataset if its parent urls has the same sample count and text length
+    Args:
+        dataset:
+
+    Returns:
+
+    """
+    # Create a dictionary to map each URL to its data
+    url_dict = {row['suburl']: row for row in dataset}
+
+    # Track URLs to remove
+    urls_to_remove = set()
+
+    # Iterate over each URL and compare it with its parent URL
+    for url, data in tqdm(url_dict.items()):
+        parent_url = find_parent_url(url)
+        if parent_url and parent_url in url_dict:
+            parent_data = url_dict[parent_url]
+            if (data['total_sample_count'] == parent_data['total_sample_count'] and
+                    data['total_text_length'] == parent_data['total_text_length']):
+                # If counts match, mark the higher-level URL for removal
```

Comment:
The term `higher-level` is confusing, it makes it seem like you would be removing the parent url.